hash_practice/exercises.py

```diff
@@ -1,29 +1,106 @@
+    # Step 1 - sort the word/string
+    # Step 2 - check if sorted word is in dict key, if not add as key
+    # and put the string as a value in a list.
+    # Step 3 - Return all values in dict in an array
+
+    # strings = ["eat","tea","tan","ate","nat","bat"]
+    if len(strings) == 0:
+        return strings
+
+    anagram_dict = {}
+    for word in strings:
+        sorted_string = "".join(sorted(word))
+        if sorted_string not in anagram_dict:
+            anagram_dict[sorted_string] = [word]
+        else:
+            anagram_dict[sorted_string].append(word)
+    return list(anagram_dict.values())
+
 
 def top_k_frequent_elements(nums, k):
     """ This method will return the k most common elements
         In the case of a tie it will select the first occuring element.
         Time Complexity: ?
         Space Complexity: ?
     """
-    pass
+    if len(nums) == 0:
+        return nums
+
+    heap = []
+    frequent_num_dict = {}
+    for num in nums:
+        if num not in frequent_num_dict:
+            frequent_num_dict[num] = 1
+        else:
+            frequent_num_dict[num] += 1
+
+    top_k_result = []
+    for num, count in frequent_num_dict.items():
+        heapq.heappush(heap, (count, num))
+        if len(heap) > k:
+            heapq.heappop(heap)
+
+    while len(heap):
+        top_k_result.append(heapq.heappop(heap)[1])
+    return top_k_result
 
 
 def valid_sudoku(table):
-    """ This method will return the true if the table is still
+    """ This method will return true if the table is still
         a valid sudoku table.
         Each element can either be a ".", or a digit 1-9
         The same digit cannot appear twice or more in the same 
         row, column or 3x3 subgrid
         Time Complexity: ?
         Space Complexity: ?
     """
-    pass
 
+    row_digits = {}
+    column_digits = {}
+    # this takes care of row and column duplication
+    for i in range(len(table)):
+        for j in range(len(table[i])):
+            value = table[i][j]
+            if value == ".":
+                continue
+            if value not in row_digits:
+                row_digits[value] = [i]
+            else:
+                if i in row_digits[value]:
+                    return False
+                else:
+                    row_digits[value].append(i)
+            if value not in column_digits:
+                column_digits[value] = [j]
+            else:
+                if j in column_digits[value]:
+                    return False
+                else:
+                    column_digits[value].append(j)
+
+    # this takes care of each 3 X 3 sub-grids
+    grid_digits = {}
+    for i in range(len(table)):
+        for j in range(len(table[i])):
+            value = table[i][j]
+            if value != ".":
+                if (i // 3, j // 3) in grid_digits:
+                    return False
+                else:
+                    grid_digits[(i // 3, j // 3)] = value

```

Comment:
Ooh, this is close to working, I think you just need to maintain a list rather than the single value! Either way, this was an optional problem, and you're close, so no worries.